plutus-metatheory/src/Signature.lagda.md

```diff
@@ -0,0 +1,170 @@
+
+The built-in compatible types are either type constants, type variables, 
+or type operators applied to built-in-compatible type.
+
+The type of built-in-compatible types (_⊢♯) is indexed by the number of 
+distinct type variables.
+```
+data _⊢♯ : ℕ → Set
+
+open import Builtin.Constant.Type ℕ (_⊢♯) using (TyCon)
+
+data _⊢♯ where
+  -- a type variable
+  ` : ∀ {n} → 
+      Fin n
+      --------
+    → n ⊢♯
+
+  -- a type constant or type operator applied to a built-in-compatible type
+  con : ∀ {n} → 
+      TyCon n
+      ------
+      → n ⊢♯
+```
+
+The list of arguments is a non-empty list of built-in compatible types.
+It is indexed by the number of distinct type variables that may appear.
+
+```
+
+Args : ℕ → Set
+Args n = List⁺ (n ⊢♯) 
+
+```
+
+## Signatures
+
+A signature is given by
+  1. The number (fv♯) of type variables that may appear
+  2. A list of arguments
+  3. A result type
+
+```
+record Sig : Set where 
+   constructor sig
+   field
+     -- number of type variables
+    fv♯ : ℕ   
+     -- list of arguments
+    args : Args fv♯
+     -- type of result
+    result : fv♯ ⊢♯
+
+```
+
+## Obtaining concrete types from signatures
+
+Signatures represent abstract types which need to be made concrete in 
+order to use them. The following modules may be instantiated to obtain 
+a function `sig2type` which converts from an abstract into a concrete type.
+
+The instantation is defined in two stages. In the first stage, the following parameters are instantiated:
+
+   1. the set `Con` over which types are indexed,
+   2. the types `Ty` (indexed over `Con`),
+   3. a function interpreting naturals as elements of `Con`, and 
+   4. a function interpreting variable indexes into types in `Ty.
+
+```
+module FromSig1 (Con : Set)(Ty : Con → Set) 
+               (nat2Con : ℕ → Con) 
+               (fin2Ty : ∀{n} → Fin n → Ty (nat2Con n))
+               where
+
+```
+
+With these parameters instantiated we can bring into scope type constants.
+  
+```
+  import Builtin.Constant.Type Con Ty as T2
+  open T2.TyCon
+```
+
+The second stage is instantiating the following parameters:
+  1. a way to insert a type constant as a type,
+  2. the constructor for function types, and
+  3. the constructor for Π types.
+```
+  module FromSig2 
+           (mkTyCon : ∀{n} → T2.TyCon (nat2Con n) → Ty (nat2Con n)) 
+           (_⇒_ : ∀{n} → Ty (nat2Con n) → Ty (nat2Con n) → Ty (nat2Con n))
+           (Π : ∀{n} → Ty (nat2Con (suc n)) → Ty (nat2Con n))
+          where
+
+    -- convert a built-in-compatible type into a Ty type
+    ♯2* : ∀{n} → n ⊢♯ → Ty (nat2Con n)
+    ♯2* (` x) = fin2Ty x
+    ♯2* (con integer) = mkTyCon T2.integer
+    ♯2* (con bytestring) = mkTyCon  T2.bytestring
+    ♯2* (con string) = mkTyCon  T2.string
+    ♯2* (con unit) = mkTyCon  T2.unit
+    ♯2* (con bool) = mkTyCon  T2.bool
+    ♯2* (con (list x)) = mkTyCon  (T2.list (♯2* x))
+    ♯2* (con (pair x y)) = mkTyCon  (T2.pair (♯2* x) (♯2* y))
```

Comment:
My latest proposal regarding polymorphic builtins amounts to supplying the interpretation function with a general type function application operator, so that this the interpretation of an instantiated polymorphic built-in type is iterated type application. This way built-in types are higher-kinded and we can avoid storing type variables inside of them in the AST meaning we won't need to handle them during type checking, type normalization, type substitution, various analysis functions etc.

Reply:
Ok. This would mean changing `TyCon` to be indexed by `Kind` (and changing all the code that uses it).